File: src/Bionic.Bridge.Capacitor.JS/src/GeolocationBridge.ts
import {CallbackID, GeolocationOptions, GeolocationPosition, Plugins} from '@capacitor/core';

const {Geolocation} = Plugins;

class Watcher {
  id: CallbackID | null = null;

  public geolocationCallback(position: GeolocationPosition, err?: any): void {
    if (err) {
      console.log(`Got position error: ${err.message ? err.message : err}`);
      return;
    }
    // @ts-ignore
    DotNet.invokeMethodAsync('BionicBridgeCapacitor', 'ReportNewPosition', this.id, new GeoPos(position));
  }
}

class GeoCords {
  latitude: number;
  longitude: number;
  accuracy: number;
  altitude?: number;
  altitudeAccuracy?: number;
  speed?: number;
  heading?: number;

  constructor(coords: any) {
    this.latitude = coords.latitude ? coords.latitude : null;
    this.longitude = coords.longitude ? coords.longitude : null;
    this.accuracy = coords.accuracy ? coords.accuracy : null;
    this.altitude = coords.altitude ? coords.altitude : null;
    this.altitudeAccuracy = coords.altitudeAcuracy ? coords.altitudeAcuracy : null;
    this.speed = coords.speed ? coords.speed : null;
    this.heading = coords.heading ? coords.heading : null;
  }
}

class GeoPos {
  timestamp: number;
  coords: GeoCords;
  constructor(position: any) {
    this.timestamp = position.timestamp ? position.timestamp : null;
    this.coords = new GeoCords(position.coords);
  }
}


interface IGeolocation {
  clearWatch(id: string): Promise<void>;
  getCurrentPosition(options?: GeolocationOptions): Promise<GeoPos>;
  watchPosition(options: GeolocationOptions): Promise<CallbackID>;
}

export class GeolocationBridge implements IGeolocation {
  private static watchers = new Map<CallbackID, Watcher>();

  public async clearWatch(id: string): Promise<void> {
    if (!Geolocation) return Promise.reject("Geolocation not available.");
    try {
      await Geolocation.clearWatch({ id });
      GeolocationBridge.watchers.delete(id);
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }

  public async getCurrentPosition(options?: GeolocationOptions): Promise<GeoPos> {
    if (!Geolocation) return Promise.reject("Geolocation not available.");
    try {
      const position = new GeoPos(await Geolocation.getCurrentPosition());
      return new Promise<GeoPos>(resolve => resolve(position));
    } catch (e) {
      return new Promise<GeoPos>((r, reject) => reject(e.message));
    }
  }

  public async watchPosition(options: GeolocationOptions): Promise<CallbackID> {
    if (!Geolocation) return Promise.reject("Geolocation not available.");
    try {
      const watcher = new Watcher();
      const id = await Geolocation.watchPosition(options, watcher.geolocationCallback.bind(watcher));
      watcher.id = id;
      GeolocationBridge.watchers.set(id, watcher);
      return new Promise<CallbackID>(resolve => resolve(id));
    } catch (e) {
      return new Promise<CallbackID>((r, reject) => reject(e.message));
    }
  }
}

File: src/Bionic.Bridge.Capacitor/GeolocationBridge.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bionic.Bridge.Capacitor {
    public static class GeolocationBridge {
        private static Dictionary<string, Action<GeolocationPosition>> watchers =
            new Dictionary<string, Action<GeolocationPosition>>();

        private const string GeolocationGetCurrentPositionFunctionName =
            "BionicBridge.Capacitor.Geolocation.getCurrentPosition";

        private const string GeolocationClearWatchFunctionName = "BionicBridge.Capacitor.Geolocation.clearWatch";
        private const string GeolocationWatchPositionFunctionName = "BionicBridge.Capacitor.Geolocation.watchPosition";

        public static Task<GeolocationPosition> GetCurrentPosition(GeolocationOptions options = null) {
            return JSRuntime.Current.InvokeAsync<GeolocationPosition>(GeolocationGetCurrentPositionFunctionName, options);
        }

        public static async Task<string>
            WatchPosition(GeolocationOptions options, Action<GeolocationPosition> callback) {
            var id = await JSRuntime.Current.InvokeAsync<string>(GeolocationWatchPositionFunctionName, options);
            watchers.Add(id, callback);
            return id;
        }

        public static Task ClearWatch(string watcherId) {
            watchers.Remove(watcherId);
            return JSRuntime.Current.InvokeAsync<object>(GeolocationClearWatchFunctionName, watcherId);
        }

        [JSInvokable]
        public static void ReportNewPosition(string watcherId, GeolocationPosition newPosition) {
            watchers[watcherId](newPosition);
        }
    }

    public class GeolocationOptions {
        public bool enableHighAccuracy;
        public long maximumAge;
        public bool requireAltitude;
        public long timeout;
    }

    public class GeolocationPosition {
        public Coordinates coords;
        public long timestamp;
    }

    public class Coordinates {
        public double latitude;
        public double longitude;
        public double accuracy;
        public double? altitude;
        public double? altitudeAccuracy;
        public double? speed;
        public double? heading;
    }
}